source/extensions/filters/http/dynamic_modules/filter.cc

```diff
@@ -71,6 +93,22 @@ FilterMetadataStatus DynamicModuleHttpFilter::encodeMetadata(MetadataMap&) {
   return FilterMetadataStatus::Continue;
 }
 
+void DynamicModuleHttpFilter::sendLocalReply(Code code, absl::string_view body,
+                                             std::function<void(ResponseHeaderMap& headers)> modify_headers,
+                                             const absl::optional<Grpc::Status::GrpcStatus> grpc_status,
+                                             absl::string_view details) {
+  if (sent_local_reply_) {
+    return;
+  }
+
+  sent_local_reply_ = true;
+  if (decoder_callbacks_) {
+    decoder_callbacks_->sendLocalReply(code, body, modify_headers, grpc_status, details);
+  } else if (encoder_callbacks_) {
+    encoder_callbacks_->sendLocalReply(code, body, modify_headers, grpc_status, details);
+  }
```

Comment:
I'm still not totally sure about this section. I have been trying to understand under what circumstances might `decoder_callbacks_` be null and `encoder_callbacks_` is not. And whether these are the only two options...

My guess is that it's for upstream-only filters. But i haven't been able to fully grok the upstream filters code yet.

Reply:
yeah I think only checking `decoder_callbacks_` and use it should be enough

Reply:
ok now my question is when `decoder_callbacks_` would be null? since it's a `Http::StreamFilter`, isn't the filter manager guaranteed to set the callbacks?

If the answer is that although this is the case for now, but we might want to allow different variants (e.g. upstream-only), then i think it doesn't hurt to keep it. in fact, it probably doesn't hurt to be defensive here anyway. I only ask because the only other place i see with this particular guard is wasm context, probably because in the wasm context it's not known whether what type of filter it is.

Reply:
after `Http::StreamFilterBase::onDestroy` callback is called, the decoder_callbacks_ should not be used and hence `decoder_callbacks_` should be set null in `onDestroy` but i didn't have that logic yet...

Reply:
so your observation is correct that it's currently always not null, so you can simply remove the if null checking for now in this PR - i will take care of that in the refactoring